src/main/java/slimeknights/tconstruct/smeltery/block/entity/module/MeltingModule.java

```diff
@@ -165,20 +165,30 @@ public void coolItem() {
    */
   @Nullable
   private IMeltingRecipe findRecipe() {
-    Level world = parent.getLevel();
+    Level world = parent.getParent().getLevel();
     if (world == null) {
       return null;
     }
 
     // first, try last recipe for the slot
     IMeltingRecipe last = lastRecipe;
     if (last != null && last.matches(this, world)) {
+      parent.setLastInventoryRecipe(lastRecipe);
       return last;
     }
+
+    // second, try the last recipe from the inventory
+    last = parent.getLastInventoryRecipe();
```

Comment:
Firstly, this is wrong. The cached last recipe should only be changed if we find a new recipe.

Secondly, this chunk of lines is really the responsibility of the parent, logic should look like
```
if last != null and last.matches(this)
  return last
IMeltingRecipe newRecipe = parent.findRecipe();
if newRecipe != null
  last = newRecipe
  return newRecipe
return null
```